src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs

```diff
@@ -0,0 +1,1200 @@
+
+namespace System.Net.Sockets
+{
+    // TODO:
+    // - ReceiveMessageFrom
+    // - Plumb status through async APIs to avoid callbacks on synchronous completion
+    //     - NOTE: this will require refactoring in the *Async APIs to accommodate the lack
+    //             of completion posting
+    // - Add support for unregistering + reregistering for events
+    //     - This will require a new state for each queue, unregistred, to track whether or
+    //       not the queue is currently registered to receive events
+
+    sealed class SocketAsyncContext
+    {
+        private abstract class AsyncOperation
+        {
+            public AsyncOperation Next;
+            public SocketError ErrorCode;
+
+            public AsyncOperation()
+            {
+                Next = this;
+            }
+
+            public abstract void Complete();
+        }
+
+        private sealed class SentinelAsyncOperation : AsyncOperation
+        {
+            public override void Complete()
+            {
+                Debug.Fail("SentinelAsyncOperation.Complete() should never be called");
+            }
+        }
+
+        private sealed class TransferOperation : AsyncOperation
+        {
+            public Action<int, byte[], int, int, SocketError> Callback;
+            public byte[] Buffer;
+            public BufferList Buffers;
+            public int BufferIndex;
+            public int Offset;
+            public int Count;
+            public int Flags;
+            public byte[] SocketAddress;
+            public int SocketAddressLen;
+            public int BytesTransferred;
+            public int ReceivedFlags;
+
+            public override void Complete()
+            {
+                Debug.Assert(Callback != null);
+
+                Callback(BytesTransferred, SocketAddress, SocketAddressLen, ReceivedFlags, ErrorCode);
+            }
+        }
+
+        private abstract class AcceptOrConnectOperation : AsyncOperation
+        {
+            public byte[] SocketAddress;
+            public int SocketAddressLen;
+        }
+
+        private sealed class AcceptOperation : AcceptOrConnectOperation
+        {
+            public Action<int, byte[], int, SocketError> Callback;
+            public int AcceptedFileDescriptor;
+
+            public override void Complete()
+            {
+                Debug.Assert(Callback != null);
+
+                Callback(AcceptedFileDescriptor, SocketAddress, SocketAddressLen, ErrorCode);
+            }
+        }
+
+        private sealed class ConnectOperation : AcceptOrConnectOperation
+        {
+            public Action<SocketError> Callback;
+
+            public override void Complete()
+            {
+                Debug.Assert(Callback != null);
+
+                Callback(ErrorCode);
+            }
+        }
+
+        enum State
+        {
+            Stopped = -1,
+            Set = 0,
+            Clear = 1,
+        }
+
+        struct Queue<TOperation>
```

Comment:
Consider using a different name, just so that folks reading the code don't confuse it with ```System.Collections.Generic.Queue<T>```